README.md
# Team Fudan 2023 Software Tool                           

![](https://static.igem.wiki/teams/4765/wiki/czy/rap-logo.png)

![](https://badgen.net/badge/platform/Linux,macOS,Windows?list=%7C)
![](https://badgen.net/static/Python/3.10/blue)
![](https://badgen.net/static/vue/2.6+/green)
![](https://badgen.net/static/license/CC%20BY%204.0/blue)
![](https://badgen.net/docker/size/mistyfield/rap/0.4.1-beta)

RAP is a quantitative pRAP system design software developed by Fudan iGEM 2023. Live demo of RAP is available. Visit our live demo [here](http://54.169.242.254:5000/)!

## Description

For more information, please visit our [wiki page](https://2023.igem.wiki/fudan/software).

### Features

- Intuitive WebUI with APIs for flexible development
- Compatible with GenBank format, and easy integration with [SnapGene(opens new window)](https://www.snapgene.com/)
- Full DBTL cycle support
- Experimentally validated results
- Up-to-date documentation and tutorial videos

## Installation

### Supported platforms

|    Windows     |  Linux   |  macOS   |
| :------------: | :------: | :------: |
| ✅(wsl2, amd64) | ✅(amd64) | ✅(amd64) |



## Install with docker(recommended)

Since the installation process requires docker, please install [docker](https://www.docker.com/) first.

```shell
docker pull mistyfield/kinetichub:v0.1.0-beta mistyfield/rap:0.4.1-beta mistyfield/parthub:0.1.0-beta
docker run -p 3306:3306 kinetichub
docker run -p 5000:5000 rap
docker run -p 7474:7474 -p 7687:7687 parthub
```

## Install with source code

This project requires the use of docker and yarn for deployment, so please install [docker](https://www.docker.com/) and [node.js](https://nodejs.org/en) with [yarn](https://yarnpkg.com/) first!

```shell
git clone https://gitlab.igem.org/2023/software-tools/fudan
# compile webUI
cd fudan/webUI
yarn install
cd ../
sh pack.sh
docker compose up -d
```

After that RAP will be running at http://127.0.0.1:5000.

## Usage

For more information on Usage, please visit our [wiki](https://2023.igem.wiki/fudan/software/) or [documentation](https://mistyfield.github.io/RAP-Docs/).


## Contributing
### Pull Requests

We actively welcome your pull requests.

1. Fork the repo and create your branch from `main`.
2. If you've added code that should be tested, add tests.
3. If you've changed APIs, update the documentation.
4. Make sure your code lints.

In order to accept your pull request, please send me an [email](mailto:20301050198@fudan.edu.cn).

### Issues

We use GitLab issues to track bugs. Please describe your issue clearly and give sufficient instructions to reproduce the issue.

## Authors and acknowledgment
### Author

- Zhiyue Chen ([@mistyfield](https://gitlab.igem.org/mistyfield))


CHANGELOG.md
# [](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.3.2-beta...v) (2023-10-12)


### Features

* add demo page ([a66f0c4](https://gitlab.igem.org/2023/software-tools/fudan/commit/a66f0c4db469d04552c5e7ba2c4ba9f0a4b816e4))
* finish stem loop design mode ([2d51fc4](https://gitlab.igem.org/2023/software-tools/fudan/commit/2d51fc4230a0d5052c6ca8ef96558898c50cffb9))
* update docker ([e54e0cd](https://gitlab.igem.org/2023/software-tools/fudan/commit/e54e0cd2948d21acc956746d2882b169dd9fe7db))
* update notebook ([edf907b](https://gitlab.igem.org/2023/software-tools/fudan/commit/edf907bbb60c534780a4d2c7dbfc6391afebdb47))
* update stem loop prediction ([fa95e0d](https://gitlab.igem.org/2023/software-tools/fudan/commit/fa95e0d8012de2c3dea2a138f8e13ef542b373f1))



## [0.3.2-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.3.1-beta...v0.3.2-beta) (2023-10-03)


### Features

* add annotation for sequence ([822b3b7](https://gitlab.igem.org/2023/software-tools/fudan/commit/822b3b7b2e4fd16fb1ce008b7455f0e90e8fd411))
* add stem loop design mode ([723cdc7](https://gitlab.igem.org/2023/software-tools/fudan/commit/723cdc795f0e67acf8068b059e01792569750ce8))
* upload data file for 2022 ([2f9d33e](https://gitlab.igem.org/2023/software-tools/fudan/commit/2f9d33e2c44e82f703abb4ecc9504b10578ce722))



## [0.3.1-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.3.0-beta...v0.3.1-beta) (2023-09-06)


### Features

* improve the colorbar and modify some ui ([ae1d8b6](https://gitlab.igem.org/2023/software-tools/fudan/commit/ae1d8b6d8e3993636cf24eeba9eae6aaa92d254f))



# [0.3.0-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.2.3-beta...v0.3.0-beta) (2023-09-06)


### Features

* add backend of PartHub ([3a4d090](https://gitlab.igem.org/2023/software-tools/fudan/commit/3a4d0908f4c854c8330ff2566fbdc0116063d03d))
* embed neovis.js for visualization ([84cf841](https://gitlab.igem.org/2023/software-tools/fudan/commit/84cf841e0aa000529217cab088d5f4f9bb9c03d4))



## [0.2.3-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.2.2-beta...v0.2.3-beta) (2023-09-03)


### Features

* finish search page of parthub2 and add backend of paryhub2 ([c831bc8](https://gitlab.igem.org/2023/software-tools/fudan/commit/c831bc8ed81c517cb1b49da34db8240904182ef2))



## [0.2.2-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.2.1-beta...v0.2.2-beta) (2023-09-01)


### Features

* add fornt end of parthub2 ([20d51d1](https://gitlab.igem.org/2023/software-tools/fudan/commit/20d51d15ee52900d0bc590e201c0dabc3bbf8a20))
* add search query highlight ([19b6e2a](https://gitlab.igem.org/2023/software-tools/fudan/commit/19b6e2a5b219829d88ecdced60ce1a33cadbd201))



## [0.2.1-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.2.0-beta...v0.2.1-beta) (2023-08-31)


### Features

* optimize the download interface and optimize the export GeneBank file format ([34ebb02](https://gitlab.igem.org/2023/software-tools/fudan/commit/34ebb025bd927bb49894b7f045f8efa221bf3596))



# [0.2.0-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.1.2-beta...v0.2.0-beta) (2023-08-30)


### Bug Fixes

* fix import error of rbsbuilder ([505c0a8](https://gitlab.igem.org/2023/software-tools/fudan/commit/505c0a897e6351684b1659bbf8186689b9f52509))


### Features

* add privacy policy ([3cc007e](https://gitlab.igem.org/2023/software-tools/fudan/commit/3cc007eb42c00d0496a973a50a1d16e5f928d2e0))
* coupling the calculated optimal ratio of enzyme concentration to rbs intensity ([c654aa4](https://gitlab.igem.org/2023/software-tools/fudan/commit/c654aa4edcd9803857131e016007b2208b2a0b35))
* finish RAP Builder ([5e58dd1](https://gitlab.igem.org/2023/software-tools/fudan/commit/5e58dd1de03c4e063294cb4acce8be3c4815f946))
* **requirements.txt:** update requirements for rapBuilder ([e077644](https://gitlab.igem.org/2023/software-tools/fudan/commit/e077644b15e5781219d1f3688e0bc4d217b8ebb9))



## [0.1.2-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.1.1-beta...v0.1.2-beta) (2023-08-25)


### Features

* create backend of RAPBuilder ([b7f22e6](https://gitlab.igem.org/2023/software-tools/fudan/commit/b7f22e6a1ce06d187dffc74463e8be5beab195ef))



## [0.1.1-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/v0.1.0-beta...v0.1.1-beta) (2023-08-23)


### Features

* beautify web view and add mirror source to accerlate docker image build ([d0b6c16](https://gitlab.igem.org/2023/software-tools/fudan/commit/d0b6c16d45a2885f619a0d5cdaba24fc0e8811eb))
* upgrade layout of index page ([c5022c7](https://gitlab.igem.org/2023/software-tools/fudan/commit/c5022c753155380f03ce939d152065442f206587))



# [0.1.0-beta](https://gitlab.igem.org/2023/software-tools/fudan/compare/423f919c132ce9ee95f0d53f2d44dd12c6b7bc34...v0.1.0-beta) (2023-08-12)


### Bug Fixes

* change .gitignore ([970e55f](https://gitlab.igem.org/2023/software-tools/fudan/commit/970e55fe0cadc86bddc477cd2764c95313c3e51a))
* fix bg.svg ([2a884c2](https://gitlab.igem.org/2023/software-tools/fudan/commit/2a884c21975c8a67544b972b10626a1f308da583))
* fix form keys in build reactions in frontend ([204209c](https://gitlab.igem.org/2023/software-tools/fudan/commit/204209c70e8b61869f8792b785bf329bc916afc5))
* lint python code ([896e8b1](https://gitlab.igem.org/2023/software-tools/fudan/commit/896e8b1d3c0f375833ed60ca0b3c551ca1f271c0))


### Features

* add add enzyme record feature ([ebe8ac8](https://gitlab.igem.org/2023/software-tools/fudan/commit/ebe8ac805da48ca931da42eacc2b0fd2921f53df))
* add build reactions for backend ([7269e0b](https://gitlab.igem.org/2023/software-tools/fudan/commit/7269e0b670d8f109cbfa5eca9a82041518312962))
* add build reactions for frontend ([d509f97](https://gitlab.igem.org/2023/software-tools/fudan/commit/d509f979a67c02b1865d7f56e3adbce4bb52ad4a))

upload_parthub2.py
"""
#!/usr/bin/env python
# -*- coding:utf-8 -*-
@File : upload_parthub2.py
@Author : Zhiyue Chen
@Time : 2023/10/12 21:10
"""
from py2neo import Node, Relationship, Graph, Subgraph
import pandas as pd
from tqdm import tqdm
import re


def gradient_color_generate(time: str):
    from datetime import datetime
    import matplotlib as mpl
    begin_time = datetime.strptime('2004-1-1', r'%Y-%m-%d')
    try:
        exact_time = datetime.strptime(time, r'%Y-%m-%d')
    except:
        exact_time = datetime.strptime('2004-1-1', r'%Y-%m-%d')
    now = datetime.now()
    max_interavl = now - begin_time
    interval = now - exact_time
    norm = mpl.colors.Normalize(vmin=0, vmax=max_interavl.days)
    cmap = mpl.colormaps.get_cmap('plasma')
    hex_rgb = mpl.colors.rgb2hex(cmap(norm(interval.days)))
    return hex_rgb


# with neo4j running
graph = Graph("bolt://parthub:7687", auth=("neo4j", "igem2023"))
graph.delete_all()
data = pd.read_csv(r'all_collections_filted.csv')
part_node_dict = {}
part_list = []
relationship_list = []
for i in tqdm(data.index):
    part_num = str(data['part_num'].values[i])
    part_name = str(data['part_name'].values[i])
    part_url = str(data['part_url'].values[i])
    part_desc = str(data['short_desc'].values[i])
    part_type = str(data['part_type'].values[i])
    part_team = str(data['team'].values[i])
    part_sequence = str(data['sequence'].values[i])
    part_contents = str(re.sub(' Sequence and Features', '', str(data['contents'].values[i])))
    part_released = str(data['released'].values[i])
    part_sample = str(data['sample'].values[i])
    part_twins = str(data['twins'].values[i])
    part_assemble = str(data['assemble_std'].values[i])
    part_used = str(data['parts_used'].values[i])
    part_using = str(data['using_parts'].values[i])
    part_len = str(data['len'].values[i])
    part_date = str(data['date'].values[i])
    part_isfavorite = str(data['isfavorite'].values[i])
    part_year = str(data['year'].values[i])
    part_designer = str(data['designer'].values[i])
    try:
        part_used_list = part_used.split(' ')
        part_using_list = part_using.split(' ')
        part_twins_list = part_twins.split(' ')
        if part_used == 'None' or part_used == '' or part_used == 'N o n e':
            part_used_list = []
        if part_using == 'self' or part_using == '':
            part_using_list = []
        if part_twins == 'None' or part_twins == '' or part_twins == 'N o n e':
            part_twins_list = []
    except:
        part_used_list = []
        part_using_list = []
        part_twins_list = []
    part_node = Node('Part', number=str(part_num), name=part_name, url=part_url, description=part_desc, type=part_type,
                     team=part_team, sequence=part_sequence, contents=part_contents, released=part_released,
                     sample=part_sample, assemble=part_assemble, length=part_len, date=part_date,
                     isfavorite=str(part_isfavorite), twins=part_twins_list, twins_num=str(len(part_twins_list)),
                     cited_by=part_used_list, year=part_year, cites=str(len(part_used_list)), ref=part_using_list,
                     citing=str(len(part_using_list)), designer=part_designer, prweight=max(1,len(part_used_list) * 0.5+len(part_using_list)+0.75 * len(part_twins_list)),
                     color=gradient_color_generate(part_date))
    part_list.append(part_node)
    part_node_dict.update({str(part_num): part_node})
twins_set_list = []
twins_node_list = []
for pNode in tqdm(part_node_dict.values()):
    if pNode['ref']:
        for ref_part in pNode['ref']:
            try:
                pNode1 = part_node_dict[ref_part]
                relationShip = Relationship(pNode, 'refers to', pNode1)
                relationShip["weight"] = pNode['prweight']
                relationship_list.append(relationShip)
            except:
                pass
    if pNode['twins']:
        for twin_part in pNode['twins']:
            try:
                pNode2 = part_node_dict[twin_part]
                if pNode2['number'] != pNode['number']:
                    if set([pNode2['number'],pNode['number']]) not in twins_set_list:
                        relationShip = Relationship(pNode, 'twins', pNode2)
                        relationship_list.append(relationShip)
                        twins_set_list.append(set([pNode2['number'],pNode['number']]))
            except:
                pass
    if pNode['cited_by']:
        for cite_part in pNode['cited_by']:
            try:
                pNode3 = part_node_dict[cite_part]
                pNode3 = part_node_dict[cite_part]
                relationShip = Relationship(pNode3, 'refers to', pNode)
                relationShip["weight"] = pNode3['prweight']
                relationship_list.append(relationShip)
            except:
                pass
relationship_list = list(set(relationship_list))
subgraph = Subgraph(part_list, relationship_list)
tx = graph.begin()
tx.create(subgraph)
graph.commit(tx)
# create graph
query = """
CALL gds.graph.project(
  'parthub',
  'Part',
  'refers to',
  {
    relationshipProperties: 'weight'
  }
)
"""
graph.run(query)
# calculate PageRanks
query = '''
CALL gds.pageRank.write('parthub', {
  maxIterations: 20,
  dampingFactor: 0.85,
  writeProperty: 'pagerank',
  relationshipWeightProperty: 'weight'
})
YIELD nodePropertiesWritten, ranIterations
'''
graph.run(query)
# get max pagerank and min pagerank
query = '''
MATCH (n:Part)
RETURN max(n.pagerank) AS max_val, min(n.pagerank) AS min_val
'''
graph.run(query)
# generate nodesize
query = '''
MATCH (n:Part)
SET n.nodesize = (n.pagerank - 0.15000000000000002) / (46.258541601845714 - 0.15000000000000002) * 90 + 30
'''
graph.run(query)
# run Louvain method
query = '''
CALL gds.louvain.write('parthub', {
  writeProperty: 'community',
  relationshipWeightProperty: 'weight'
})
'''
graph.run(query)
query = '''
CALL gds.graph.drop('parthub')
'''
graph.run(query)